src/Elliptic/Actions/InitializeAnalyticSolution.hpp

```diff
@@ -32,47 +32,52 @@ namespace Actions {
 /*!
  * \brief Place the analytic solution of the system fields in the DataBox.
  *
+ * If the `BackgroundTag` holds an `AnalyticSolutionType`, evaluate it for all
+ * `AnalyticSolutionFields` and store the fields in the DataBox. Otherwise,
+ * signal no analytic solution is available by storing `std::nullopt` in the
+ * DataBox.
+ *
  * Uses:
  * - DataBox:
- *   - `AnalyticSolutionTag`
+ *   - `BackgroundTag`
  *   - `Tags::Coordinates<Dim, Frame::Inertial>`
  *
  * DataBox:
  * - Adds:
- *   - `db::wrap_tags_in<::Tags::Analytic, AnalyticSolutionFields>`
+ *   - `::Tags::AnalyticSolutions<AnalyticSolutionFields>`
  *
  * \note This action relies on the `SetupDataBox` aggregated initialization
  * mechanism, so `Actions::SetupDataBox` must be present in the `Initialization`
  * phase action list prior to this action.
  */
-template <typename AnalyticSolutionTag, typename AnalyticSolutionFields>
+template <typename BackgroundTag, typename AnalyticSolutionFields,
+          typename AnalyticSolutionType>
 struct InitializeAnalyticSolution {
-  using const_global_cache_tags = tmpl::list<AnalyticSolutionTag>;
-  using analytic_fields_tag =
-      db::add_tag_prefix<::Tags::Analytic,
-                          ::Tags::Variables<AnalyticSolutionFields>>;
+  using analytic_fields_tag = ::Tags::AnalyticSolutions<AnalyticSolutionFields>;
 
   using simple_tags = tmpl::list<analytic_fields_tag>;
   using compute_tags = tmpl::list<>;
 
   template <typename DbTagsList, typename... InboxTags, typename Metavariables,
             size_t Dim, typename ActionList, typename ParallelComponent>
-  static auto apply(db::DataBox<DbTagsList>& box,
-                    const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
-                    const Parallel::GlobalCache<Metavariables>& cache,
-                    const ElementId<Dim>& /*array_index*/,
-                    const ActionList /*meta*/,
-                    const ParallelComponent* const /*meta*/) noexcept {
-
-    const auto& inertial_coords =
-        get<domain::Tags::Coordinates<Dim, Frame::Inertial>>(box);
-    typename analytic_fields_tag::type analytic_fields{
-        variables_from_tagged_tuple(get<AnalyticSolutionTag>(cache).variables(
-            inertial_coords, AnalyticSolutionFields{}))};
-    Initialization::mutate_assign<simple_tags>(make_not_null(&box),
-                                               std::move(analytic_fields));
-
-    return std::make_tuple(std::move(box));
+  static std::tuple<db::DataBox<DbTagsList>&&> apply(
+      db::DataBox<DbTagsList>& box,
+      const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      const Parallel::GlobalCache<Metavariables>& /*cache*/,
+      const ElementId<Dim>& /*array_index*/, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    const auto analytic_solution =
```

Comment:
Sorry, I'm now confused about how this dynamic cast is supposed to work in the generic case. If I'm reading it correctly, you're anticipating that the `BackgroundTag` will hold some factory creatable thing, that may or may not be a derived class of `AnalyticSolutionType`. But then it seems like you have `SomeClass` that inherits both from the base class in `BackgroundTag` and from `AnalyticSolutionType` and you are trying to cast between the two base classes, but I don't think that cast works when there isn't a direct inheritance.

And if there is a direct inheritance, then I think you should be able to just use `std::is_base_of` and avoid the dynamic cast trickery.

Is there a case here that I'm missing?

Reply:
The inheritance I was aiming for is `SomeClass : AnalyticSolution : Background : AnalyticData` for analytic solutions and `SomeClass : Background : AnalyticData` for analytic data, where the `Background` class may be omitted for systems that don't have extra requirements like a constitutive relation or an EOS.

The `BackgroundTag::type` here is `Background` and the `AnalyticSolutionType` is `AnalyticSolution`. So I don't think `std::is_base_of` would work because I need to determine at runtime which of the two inheritance trees the object is part of.